src/core_deku/ticket_table.ml

```diff
@@ -0,0 +1,126 @@
+open Helpers
+
+module Tickets = struct
+  module Ticket_map = Map.Make_with_yojson (Ticket_id)
+
+  type t = Amount.t Ticket_map.t [@@deriving yojson]
+
+  let get_balance t ~ticket = Ticket_map.find_opt ticket t
+
+  let of_seq = Ticket_map.of_seq
+
+  let singleton ~ticket ~amount =
+    Ticket_map.of_seq (Seq.return (ticket, amount))
+
+  let merge t ~ticket ~amount =
+    let merger option ~amount =
+      let value = Option.fold ~none:amount ~some:Amount.(( + ) amount) option in
+      value in
+    Ticket_map.update ticket
+      (fun x ->
+        let merged = merger ~amount x in
+        Option.some merged)
+      t
+
+  let get_and_remove t ~ticket ~to_take =
+    let%ok amount =
+      Ticket_map.find_opt ticket t |> Option.to_result ~none:`Insufficient_funds
+    in
+    let%assert () =
+      ( `Insufficient_funds,
+        let compared = Amount.compare amount to_take in
+        compared >= 0 ) in
+    let map =
+      Ticket_map.update ticket
+        (function
+          | Some amount ->
+            if Amount.equal amount to_take then
+              None
+            else
+              Some Amount.(amount - to_take)
+          | None -> assert false)
+        t in
+    Ok map
+
+  let subtract t ~ticket ~to_withdraw =
+    let%ok map = get_and_remove t ~ticket ~to_take:to_withdraw in
+    Ok map
+
+  let fetch t tickets =
```

Comment:
Let's try to have more consistent names. We could rename this to `get_and_remove_many`.